builder/cpp_builder.rb

```diff
@@ -304,33 +305,58 @@ def file_to_test(filename)
   end
 
   def run_tests
-    xml_log = "#{@abs_cpp_test_out_dir}/results.xml"
+    attempt = 1
```

Comment:
This solution adds yet another trying loop with its own attempts counter, etc.

As alternative idea: will it bring any benefits we detect offending (aborted) test and then feed that information back to normal partial builder mechanism to exclude offending files out of the binary to get us the stable set of tests in one binary in the end?

On one hand:

* this avoids looking into boost::test invocation flags which might be not compatible with boost versions you've mentioned
* this requires test parsing to determine offending test, but you did that already

On the other hand, I see that such a feedback mechanism from `run_tests` does not exist now and inventing it on whole `partial_builder.rb` might be trickier...

Reply:
> This solution adds yet another trying loop with its own attempts counter, etc.

Yeah. I don't think you can eliminate it completely in principle, because you don't know in advance which tests will abort, and as soon one test aborts, no more tests are run (this is by design, Boost has no "continue running the suite despite an aborted test" option, that's not a thing). If you want them to run, you have no option but to run the test executable again without that offending test.

> As alternative idea: will it bring any benefits we detect offending (aborted) test and then feed that information back to normal partial builder mechanism to exclude offending files out of the binary to get us the stable set of tests in one binary in the end?

I briefly considered this approach as well, but don't see any benefits. In my view, it would only be more complicated and less flexible. As I explain above, you have to run the test suite `N + 1` times in any case (where `N` is the number of aborted tests), only now with the unnecessary extra step of repeating the linking the object files into the `ks_tests` executable.

> * this avoids looking into boost::test invocation flags which might be not compatible with boost versions you've mentioned

This is no longer a problem because since https://github.com/kaitai-io/kaitai_struct_docker_images/commit/1ca846db680421fde3f6ad5bdff62b61bac5ed11 the `{clang3.4,gcc4.8}-linux-x86_64` targets build the latest Boost 1.84 from source (instead of using ancient the Boost 1.54), which works like a charm. None of Boost versions we use at the moment have a problem with `--run_test=!<test_name>` anymore (as evidenced by https://ci.kaitai.io/ - coincidentally, all `cpp_stl_{98,11}` targets currently have at least one aborted test `NavParentRecursive`, yet they all recovered successfully), so there's no need to avoid it.

---

An actual disadvantage of the approach you're suggesting is that with the "stable" test binary at the end, you won't be able to run the aborted tests via Valgrind after `CppBuilder` has finished. This means that the only information about an aborted will be the message from when the OS signal was caught by Boost, i.e.:

```
memory access violation at address: 0x00000008: no mapping at fault address
```

... which isn't really helpful to diagnose the error - there's no file name, no line number, no stack trace. Valgrind can give you all that, if you only let it run the aborted test (`cpp_stl_11/gcc11-linux-x86_64` > [test_out/cpp_stl_11/valgrind-1.xml:36-57](https://github.com/kaitai-io/ci_artifacts/blob/4ee6502cedb050b22cd0c9cde009006947c53fa8/test_out/cpp_stl_11/valgrind-1.xml#L36-L57)):

```xml
<error>
  <unique>0x0</unique>
  <tid>1</tid>
  <kind>InvalidRead</kind>
  <what>Invalid read of size 8</what>
  <stack>
    <frame>
      <ip>0x54235A</ip>
      <obj>/tests/compiled/cpp_stl_11/bin/ks_tests</obj>
      <fn>debug_array_user_t::cat_t::_read()</fn>
      <dir>/tests/compiled/cpp_stl_11</dir>
      <file>debug_array_user.cpp</file>
      <line>37</line>
    </frame>
    <frame>
      <ip>0x542205</ip>
      <obj>/tests/compiled/cpp_stl_11/bin/ks_tests</obj>
      <fn>debug_array_user_t::_read()</fn>
      <dir>/tests/compiled/cpp_stl_11</dir>
      <file>debug_array_user.cpp</file>
      <line>20</line>
    </frame>
```

(Valgrind logs of aborted tests are currently produced, but not processed by `aggregate/convert_to_json` - I left this as a future improvement.)